Code/GraphMol/ForceFieldHelpers/UFF/Builder.cpp

```diff
@@ -84,48 +91,39 @@ boost::uint8_t getTwoBitCell(boost::shared_array<boost::uint8_t> &res,
 // ------------------------------------------------------------------------
 boost::shared_array<boost::uint8_t> buildNeighborMatrix(const ROMol &mol) {
   unsigned int nAtoms = mol.getNumAtoms();
-  unsigned nTwoBitCells = (nAtoms * nAtoms - 1) / 4 + 1;
+  unsigned nTwoBitCells = (nAtoms * (nAtoms + 1) - 1) / 8 + 1;
   boost::shared_array<boost::uint8_t> res(new boost::uint8_t[nTwoBitCells]);
-  for (unsigned int i = 0; i < nTwoBitCells; ++i) {
-    res[i] = 0;
-  }
   for (unsigned int i = 0; i < nAtoms; ++i) {
-    unsigned int iTab = i * nAtoms;
     for (unsigned int j = i; j < nAtoms; ++j) {
-      setTwoBitCell(res, iTab + j, RELATION_1_X);
-      setTwoBitCell(res, i + j * nAtoms, RELATION_1_X);
+      setTwoBitCell(res, twoBitCellPos(nAtoms, i, j), RELATION_1_X);
     }
   }
-  for (unsigned int i = 0; i < mol.getNumBonds(); ++i) {
-    const Bond *bondi = mol.getBondWithIdx(i);
-
-    setTwoBitCell(res,
-                  bondi->getBeginAtomIdx() * nAtoms + bondi->getEndAtomIdx(),
-                  RELATION_1_2);
-    setTwoBitCell(res,
-                  bondi->getEndAtomIdx() * nAtoms + bondi->getBeginAtomIdx(),
-                  RELATION_1_2);
-
-    for (unsigned int j = i + 1; j < mol.getNumBonds(); ++j) {
-      const Bond *bondj = mol.getBondWithIdx(j);
+  for (ROMol::ConstBondIterator bondi = mol.beginBonds(); bondi != mol.endBonds(); ++bondi) {
+    setTwoBitCell(res, twoBitCellPos(nAtoms, (*bondi)->getBeginAtomIdx(),
+                  (*bondi)->getEndAtomIdx()), RELATION_1_2);
+    for (ROMol::ConstBondIterator bondj = bondi; bondj != mol.endBonds(); ++bondj) {
+      if (*bondj == *bondi) continue;
       int idx1 = -1;
       int idx3 = -1;
-      if (bondi->getBeginAtomIdx() == bondj->getBeginAtomIdx()) {
-        idx1 = bondi->getEndAtomIdx();
-        idx3 = bondj->getEndAtomIdx();
-      } else if (bondi->getBeginAtomIdx() == bondj->getEndAtomIdx()) {
-        idx1 = bondi->getEndAtomIdx();
-        idx3 = bondj->getBeginAtomIdx();
-      } else if (bondi->getEndAtomIdx() == bondj->getBeginAtomIdx()) {
-        idx1 = bondi->getBeginAtomIdx();
-        idx3 = bondj->getEndAtomIdx();
-      } else if (bondi->getEndAtomIdx() == bondj->getEndAtomIdx()) {
-        idx1 = bondi->getBeginAtomIdx();
-        idx3 = bondj->getBeginAtomIdx();
+      unsigned int bondiBeginAtomIdx = (*bondi)->getBeginAtomIdx();
+      unsigned int bondiEndAtomIdx = (*bondi)->getEndAtomIdx();
```

Comment:
Note, bondi doesn't change inside this inner loop, so setting bondiBeginAtomIdx and bondiEndAtomIdx can be done once for this whole inner loop. So if this refactoring speeded things up, could push it a little further hopefully.